---
title: A pipeline in R for the analysis of Infinium methylation array data
author: Saul Pierotti
date: June 14, 2020
output: pdf_document
---

This report was initially written as a Jupyter notebook, then migrated to Rmarkdown for compatibility issues with some packages.
This report serves as a final assignment for the DNA and RNA dynamics course of the Bioinformatics master degree of the University of Bologna (AA 2019/2020).
The student specific parameters to be used for the report (as detailed in the assignment) are reported below.

| Parameter                |               Value |
|:-------------------------|--------------------:|
| Student number           |                  24 |
| Step 3 address           |            59625465 |
| Step 5 p-value threshold |                0.01 |
| Step 7 normalization     |   preprocessFunnorm |
| Step 9 test              |   Mann-Whitney test |

# Step 1
>Load raw data with minfi and create an object called RGSet storing the RGChannelSet object.

I first clean the workspace and then load the required libraries. I do not set the working directory since I will give full paths relative to the directory in which this notebook is run.


```{r, message=FALSE, warning=FALSE}
rm(list=ls())
```


```{r, message=FALSE, warning=FALSE}
require('minfi')
require(IlluminaHumanMethylation450kmanifest)
require(gplots)
require(gap)
require(amap)
data_dir <- './Input_data'
```

I load the sample sheet of the experiment using the minfi `read.metharray.sheet` function.


```{r, message=FALSE, warning=FALSE}
targets <- read.metharray.sheet(data_dir)
```

Now I load the raw experimental data relative to the sample sheet and I create the RGChannelSet object using the `read.metharray.exp` function.


```{r, message=FALSE, warning=FALSE}
RGSet <- suppressWarnings(read.metharray.exp(base='./Input_data',targets = targets))
dim(RGSet)
```

# Step 2
>Create the dataframes Red and Green to store the red and green fluorescences respectively.

I use the `getGreen` and `getRed` functions on the RGSet for extracting data from the respective color channels.
These data are then converted to 2 data frames of the same dimensionality (8\*622399, 8 samples and 622399 probe addresses).


```{r, message=FALSE, warning=FALSE}
Red <- data.frame(getRed(RGSet))
Green <- data.frame(getGreen(RGSet))
dim(Red)
dim(Green)
```

# Step 3
> Fill the following table: what are the Red and Green fluorescences for the address assigned to you?
> Optional: check in the manifest file if the address corresponds to a Type I or a Type II probe and, in case of Type I probe, report its color.

The student-specific probe address assigned for this step is `59625465`.
First I note down which array and slide refers to which sample by reading the sample sheet.


```{r, message=FALSE, warning=FALSE}
targets
```

Then I extract the red and green intensities for the assigned probe by slicing the respective data frames (Red and Green).


```{r, message=FALSE, warning=FALSE}
Red[rownames(Red)=="59625465",]
Green[rownames(Green)=="59625465",]
```

Now I load the manifest for the Illumina 450k Infinium array with the `getProbeInfo` function and check what is the chemistry of the assigned probe.

```{r, message=FALSE, warning=FALSE}
probe_info_I <- getProbeInfo(RGSet, type = 'I')
probe_info_II <- getProbeInfo(RGSet, type = 'II')
probe_info_I[probe_info_I$AddressA==59625465,]
probe_info_I[probe_info_I$AddressB==59625465,]
probe_info_II[probe_info_II$AddressA==59625465,]
```

I find that it is a type II probe, so there is no color to specify in the table.
With all of these pieces of information I can complete the table as assigned (I also included addition information for clarity and completeness).


| Sample | Group |      Slide | Array  | Red fluorescence | Green fluorescence | Type | Color |
|--------|-------|-----------:|--------|-----------------:|-------------------:|-----:|------:|
| 1020   | DS    | 5775278051 | R01C01 |            11585 |                935 |   II |     / |
| 1036   | DS    | 5775278051 | R04C02 |            14282 |                845 |   II |     / |
| 3038   | WT    | 5775278078 | R02C01 |            11253 |                666 |   II |     / |
| 3042   | WT    | 5775278078 | R05C01 |            11494 |                785 |   II |     / |
| 3052   | WT    | 5775278078 | R05C02 |            11254 |                652 |   II |     / |
| 1016   | DS    | 5930514034 | R01C02 |            11152 |                306 |   II |     / |
| 1029   | DS    | 5930514035 | R04C02 |            11625 |                584 |   II |     / |
| 3029   | WT    | 5930514035 | R06C02 |            13019 |                668 |   II |     / |

# Step 4
> Create the object MSet.raw

I use the `preprocessRaw` function on the RGSet to get the raw MSet.


```{r, message=FALSE, warning=FALSE}
MSet.raw <- preprocessRaw(RGSet)
dim(MSet.raw)
```

# Step 5
> Perform the following quality checks and provide a brief comment to each step:
> * QCplot
> * check the intensity of negative controls using minfi
> * calculate detection pValues; for each sample, how many probes have a detection p-value higher than the threshold assigned to each student?

I first create a QC plot by getting the medians of the methylated and unmethylated channels for each sample with the `getQC` function and plotting them with the `plotQC` function.


```{r, message=FALSE, warning=FALSE}
qc <- getQC(MSet.raw)
plotQC(qc)
```

Since all the samples have high median methylated and unmethylated signals, I can consider that all of them are of good quality.
Now I plot the negative control probe intensities of the RGSet using the `controlStripPlot` function.


```{r, message=FALSE, warning=FALSE}
controlStripPlot(RGSet, controls="NEGATIVE")
```

The represented color are inverted (the red is shown in green and vice-versa). This was observed also during the lectures and it can be a bug of the package (minfi). During the lectures we also determined that the text label refers to the correct color, not the actual color of the plot.
The negative controls are all fine since their intensity is at background levels (below 10 in Log2 scale).
We can notice how the negative control probes intensities show slight differences across slides in a manner that is fairly consistent across the 2 color channels.

Now I will determine the detection p-values for the probes using the `detectionP` function on the RGSet.
I will consider as failed the probes with a detection p-value higher than the assigned student-specific threshold of 0.01.


```{r, message=FALSE, warning=FALSE}
detP <- detectionP(RGSet)
dim(detP)
failed <- detP>0.01
table(failed)
summary(failed)
```
The following number of probes has a p-value higher than the threshold for each one of the samples:

| Sample | Group |      Slide | Array  | Failed probes (p-value > 0.01) |
|--------|-------|-----------:|--------|-------------------------------:|
| 1020   | DS    | 5775278051 | R01C01 |                            323 |
| 1036   | DS    | 5775278051 | R04C02 |                            260 |
| 3038   | WT    | 5775278078 | R02C01 |                            312 |
| 3042   | WT    | 5775278078 | R05C01 |                            485 |
| 3052   | WT    | 5775278078 | R05C02 |                            465 |
| 1016   | DS    | 5930514034 | R01C02 |                            123 |
| 1029   | DS    | 5930514035 | R04C02 |                             60 |
| 3029   | WT    | 5930514035 | R06C02 |                            149 |

Even if not specified in the assignment, it can be useful to look for the percentages of failed probes in each sample.

```{r, message=FALSE, warning=FALSE}
sample_failed_probes_percentage <- colMeans(failed)
sample_failed_probes_percentage
```
All the samples seem to show an acceptable amount of failed probes.
I can also check for probes that failed (according to the aforementioned criterion) on more than 1% of the samples (also this was not required in the assignment).

```{r, message=FALSE, warning=FALSE}
mean_failure_probes <- rowMeans(failed)
probes_to_be_retained <- mean_failure_probes<0.01
summary(probes_to_be_retained)
probes_to_be_removed <- !probes_to_be_retained
names_probes_to_be_removed <- names(probes_to_be_removed)
```
There are 1307 probes that failed in more than 1% of the samples.
I will save their name in a vector to eventually filter them out later, but I will retain them for now so to avoid possible problems with processing due to missing probes.

# Step 6
> Calculate raw beta and M values and plot the densities of mean methylation values, dividing the samples in DS and WT (suggestion: subset the beta and M values matrixes in order to retain DS or WT subjects and apply the function mean to the 2 subsets). 

I first extract the beta and M values from the MSet.raw using the `getBeta` and `getM` functions.


```{r, message=FALSE, warning=FALSE}
beta <- getBeta(MSet.raw)
summary(beta)
M <- getM(MSet.raw)
summary(M)
```

Now I subset the beta and M value matrices on the basis of the group (DS or WT) of the sample.


```{r, message=FALSE, warning=FALSE}
isDS <- targets$Group=='DS'
summary(isDS)
beta_DS <- beta[,isDS]
beta_WT <- beta[,!isDS]
M_DS <- M[,isDS]
M_WT <- M[,!isDS]
```

Now I calculate the mean beta and M values for each queried position across the 2 sample groups (across rows, MARGIN=1), discarding NA values (na.rm=T).
I need to discard NA values since, where both the methylated and unmethilated signals are 0, the beta and M values are not defined.


```{r, message=FALSE, warning=FALSE}
beta_DS_mean <- apply(beta_DS,MARGIN=1,mean,na.rm=T)
beta_WT_mean <- apply(beta_WT,MARGIN=1,mean,na.rm=T)
M_DS_mean <- apply(M_DS,MARGIN=1,mean,na.rm=T)
M_WT_mean <- apply(M_WT,MARGIN=1,mean,na.rm=T)
```

Now I can calculate the density distribution of all of these means and plot them.


```{r, message=FALSE, warning=FALSE}
beta_DS_mean_d <- density(beta_DS_mean)
beta_WT_mean_d <- density(beta_WT_mean)
M_DS_mean_d <- density(M_DS_mean)
M_WT_mean_d <- density(M_WT_mean)
```


```{r, message=FALSE, warning=FALSE}
par(mfrow=c(1,2))
plot(beta_DS_mean_d,main="Density of Beta Values",col="red")
lines(beta_WT_mean_d,main="Density of Beta Values",col="blue")
plot(M_DS_mean_d,main="Density of M Values",col="red")
lines(M_WT_mean_d,main="Density of M Values",col="blue")
```

The red lines represents the DS samples beta/M mean densities while the blue line represents that of the WT samples.
We can see how in the 2 sample groups both beta and M values tend to have a very similar distribution.

# Step 7
> Normalize the data using the function assigned to each student and compare raw data and normalized data.
> Produce a plot with 6 panels in which, for both raw and normalized data, you show the density plots of beta mean values according to the chemistry of the probes, the density plot of beta standard deviation values according to the chemistry of the probes and the boxplot of beta values.
> Provide a short comment regarding the changes you observe.

I first extract the type I and type II probe names from the manifest of the array and use them to subset the beta values matrix.


```{r, message=FALSE, warning=FALSE}
type_I <- getProbeInfo(MSet.raw, type = 'I')$Name
type_II <- getProbeInfo(MSet.raw, type = 'II')$Name
beta_I <- beta[rownames(beta) %in% type_I,]
beta_II <- beta[rownames(beta) %in% type_II,]
dim(beta_I)
dim(beta_II)
```

I then calculate mean and standard deviation densities for type I and type II probes.


```{r, message=FALSE, warning=FALSE}
mean_beta_I <- apply(beta_I,1,mean,na.rm=T)
mean_beta_II <- apply(beta_II,1,mean,na.rm=T)
d_mean_beta_I <- density(mean_beta_I)
d_mean_beta_II <- density(mean_beta_II)
sd_beta_I <- apply(beta_I,1,sd,na.rm=T)
sd_beta_II <- apply(beta_II,1,sd,na.rm=T)
d_sd_beta_I <- density(sd_beta_I)
d_sd_beta_II <- density(sd_beta_II)
```

I normalize (between-array) the RGSet using the function assigned (`preprocessFunnorm`), which removes the variability explained by the control probes.
This will produce a GenomicRatioSet object.


```{r, message=FALSE, warning=FALSE}
normalized_data <- preprocessFunnorm(RGSet)
normalized_data
```

I can now extract the normalized beta values using the `getBeta` function and then separate type I and type II probes as done with the original data.
I also calculate mean and standard deviation densities as before.


```{r, message=FALSE, warning=FALSE}
beta_norm <- getBeta(normalized_data)
beta_I_norm <- beta[rownames(beta_norm) %in% type_I,]
beta_II_norm <- beta[rownames(beta_norm) %in% type_II,]
mean_beta_I_norm <- apply(beta_I_norm,1,mean,na.rm=T)
mean_beta_II_norm <- apply(beta_II_norm,1,mean,na.rm=T)
d_mean_beta_I_norm <- density(mean_beta_I_norm)
d_mean_beta_II_norm <- density(mean_beta_II_norm)
sd_beta_I_norm <- apply(beta_I_norm,1,sd,na.rm=T)
sd_beta_II_norm <- apply(beta_II_norm,1,sd,na.rm=T)
d_sd_beta_I_norm <- density(sd_beta_I_norm)
d_sd_beta_II_norm <- density(sd_beta_II_norm)
```

Finally I can produce plots of mean beta value densities and standard deviation densities, differentiated by chemistry of the probes and for raw and normalized data.
I also produce a box plot of the beta values across samples for the raw and normalized data.
I set the axes limits so to make the plots easier to compare.


```{r, message=FALSE, warning=FALSE}
par(mfrow=c(2,3))
plot(d_mean_beta_I,col="blue",main="raw beta",xlim=c(0,1),ylim=c(0,5))
lines(d_mean_beta_II,col="red")
plot(d_sd_beta_I,col="blue",main="raw sd",xlim=c(0,0.6),ylim=c(0,60))
lines(d_sd_beta_II,col="red")
boxplot(beta,ylim=c(0,1))
plot(d_mean_beta_I_norm,col="blue",main="preprocessFunnorm beta",xlim=c(0,1),ylim=c(0,5))
lines(d_mean_beta_II_norm,col="red")
plot(d_sd_beta_I_norm,col="blue",main="preprocessFunnorm sd",xlim=c(0,0.6),ylim=c(0,60))
lines(d_sd_beta_II_norm,col="red")
boxplot(beta_norm,ylim=c(0,1))
```

I can observe that the distribution of beta values is almost identical across samples for the normalized data (box plot), while it is more variable for the raw data. Only the median in the normalized data seems to differ slightly among samples.
The distribution of mean densities and standard deviation densities is almost equivalent among type I and II probes in the normalized data, while it is heavily different in the raw data. In the raw data the density of standard deviations for type II probes tend to be shifted towards higher values compared to type I probes.
The peaks on the mean density distribution for the type II probes are more shifted towards the center compared to type I probes in the raw data.
All of these differences among chemistry types in the raw data are as expected: type II probes are more variable and they show a narrower range of beta values compared to type I probes.

# Step 8
> Perform a PCA on the beta matrix generated in step 7.
> Comment the plot.

I want now to apply the `prcomp` function on the normalized beta values matrix obtained in step 7 in order to perform a PCA on it.
The beta values matrix has one column per sample and one probe per row, and because of this it must be transposed before applying the PCA (`t` function).


```{r, message=FALSE, warning=FALSE}
pca_results <- prcomp(t(beta_norm),scale=T)
summary(pca_results)
plot(pca_results)
```

In the screw plot, it seems that the variance is quite uniformly distributed among the first 7 components.
The 8th component contains really limited variance, and so the first 7 components can explain almost all of the variability of the dataset.
This may mean that the samples differ sensibly on more than a few orthogonal axes.
I can plot the first 2 to components to see if the DS and WT groups are separated according to those dimensions.
I can also color the same plot according to the array used, to spot batch effects.


```{r, message=FALSE, warning=FALSE}
palette("default")
groups <- factor(targets$Group)
plot(pca_results$x[,1], pca_results$x[,2],pch=16,col=groups,xlab="PC1",ylab="PC2")
legend("bottomright",legend=levels(groups),col=c(1:nlevels(groups)),pch=16)
```

I can color the same plot according to the array used, to spot batch effects.


```{r, message=FALSE, warning=FALSE}
slides <- factor(targets$Slide)
plot(pca_results$x[,1], pca_results$x[,2],pch=16,col=slides,xlab="PC1",ylab="PC2")
legend("bottomright",legend=levels(slides),col=c(1:nlevels(slides)),pch=16)
```

From these plots I can see how the 2 groups (DS and WT) seem distinct according to the first 2 principal components.
A line of the type y=x seem to well separate the 2 groups.
DS seems confined to a region where PC1>PC2, while WT is confined in a region where PC2>PC1.
The samples seem to be highly clustered according to the slide in which they were run (presumably due to batch effects).
However, slide 5930514035, which contains a sample from the DS and a sample from the WT group, still respects the pattern of separation between sample groups.
Randomization of the samples on the slides would have been advisable to allow to test for batch effects separately from sample groups.
There seem to be no outliers among the samples.

# Step 9
> Using the matrix of normalized beta values generated in step 7, identify  differentially  methylated  probes  between  group  DS  and group WT using the functions assigned to each student.
> Note; it can take several minutes; if you encounter any problem you can run the differential  methylated  analysis  only  on  a  subset  of  probes  (for example those on chromosome 1, 18 and 21)

The test assigned to me for the analysis of differentially methylated probes is the Mann-Whitney U test (Wilcoxon rank sum test), which is implemented in the `wilcox.test` function.
I apply it to the normalized beta values from step 7, using DS and WT as sample groups.
I need to define a function so to apply the test to each probe.


```{r, message=FALSE, warning=FALSE}
mann_whitney_all_rows <- function(x) {
  wilcox <- wilcox.test(x~ targets$Group)
  #wilcox <- wilcox.test(x~ groups)
  return(wilcox$p.value)
}

pval_raw <- apply(beta_norm,1, mann_whitney_all_rows)
summary(pval_raw)
length(pval_raw)
```
I create then a data frame containing the beta values and the raw p-values for each probe and sort it from smallest to largest p-value.

```{r, message=FALSE, warning=FALSE}
beta_rawp_df <- data.frame(beta_norm, pval_raw)
beta_rawp_df_sorted <- beta_rawp_df[order(beta_rawp_df$pval_raw),]
summary(beta_rawp_df_sorted)
```

# Step 10
> Apply multiple test correction and set a significant threshold of 0.05.
> How many probes do you identify as differentially methylated considering nominal pValues?
> How many after Bonferroni correction?
> How many after BH correction?

For doing multiple test correction, I will use the `p.adjust` function, specifying which correction I want to use (BH and Bonferroni).
I will apply it to the sorted raw p-values sliced from the data frame of step 9.

```{r, message=FALSE, warning=FALSE}
raw_p_sorted <- beta_rawp_df_sorted[,9]
summary(raw_p_sorted)
BH_p <- p.adjust(raw_p_sorted,"BH")
summary(BH_p)
Bonf_p <- p.adjust(raw_p_sorted,"bonferroni")
summary(Bonf_p)
```
We can note how the Bonferroni correction is too stringent, bringing all the p-values to 1.
Also with the BH correction the smallest p-value is only 0.47, so definitively not significant.
I can now count how many probes have a p-value below the threshold of 0.05.

```{r, message=FALSE, warning=FALSE}
diff_met_raw_p <- raw_p_sorted[raw_p_sorted<=0.05]
length(diff_met_raw_p)
summary(diff_met_raw_p)
diff_met_BH <- BH_p[BH_p<=0.05]
length(diff_met_BH)
diff_met_Bonf <- Bonf_p[Bonf_p<=0.05]
length(diff_met_Bonf)
```
There are no significant probes after multiple test correction.
29388 probes have a raw p-value under 0.05.
I can create a data frame including the beta values, raw p-values and corrected p-values for all the probes.
I will also create a boxplot of the raw and corrected p-values so to look at their distribution.

```{r, message=FALSE, warning=FALSE}
complete_beta_p_df <- data.frame(beta_rawp_df_sorted, BH_p, Bonf_p)
boxplot(complete_beta_p_df[,9:11])
```

# Step 11
These are the instructions:
> Produce  an  heatmap  of  the  top  100  differentially  mehtylated probes.

I will use the function `heatmap.2` to produce an heatmap of the top 100 differentially methylated probes (according to raw p-value).
I first produce an input matrix for the heatmap from the top 100 values of the sorted data frame from step 10.
I also slice the data frame so to exclude p-values and retain only beta-values.


```{r, message=FALSE, warning=FALSE}
input_heatmap <- as.matrix(complete_beta_p_df[1:100,1:8])
summary(input_heatmap)
```

I will now produce the heatmap from these top 100 probes, coloring the samples according to the group to which they belong (DS or WT).
I leave the default parameters for clustering (euclidean distance and complete linkage).


```{r, message=FALSE, warning=FALSE}
DS_col <- 'orange'
WT_col <- 'green'
colorbar <- c(DS_col, DS_col, WT_col, WT_col, WT_col, DS_col, DS_col, WT_col)
heatmap.2(input_heatmap,col=terrain.colors(100),Rowv=T,Colv=T,dendrogram="both",
          key=T,ColSideColors=colorbar,density.info="none",trace="none",scale="none",symm=F)
```

From the heatmap we notice how the 2 sample groups are well clustered according to the methylation status of the top 100 probes.
Some probes are more methylated in one of the groups, other probes in the other.
I can repeat the heatmap using single and average linkage.

```{r, message=FALSE, warning=FALSE}
heatmap.2(input_heatmap,col=terrain.colors(100),Rowv=T,Colv=T,hclustfun = function(x) hclust(x,method = 'single'),dendrogram="both",key=T,ColSideColors=colorbar,density.info="none",trace="none",scale="none",symm=F)
```
```{r, message=FALSE, warning=FALSE}
heatmap.2(input_heatmap,col=terrain.colors(100),Rowv=T,Colv=T,hclustfun = function(x) hclust(x,method = 'average'),dendrogram="both",key=T,ColSideColors=colorbar,density.info="none",trace="none",scale="none",symm=F)
```
I will also try to use Pearson and Spearman correlation as a distance metric, instead of euclidean distance.
```{r, message=FALSE, warning=FALSE}
heatmap.2(input_heatmap,col=terrain.colors(100),Rowv=T,Colv=T,distfun = function(x) Dist(x, method="pearson"),dendrogram="both",key=T,ColSideColors=colorbar,density.info="none",trace="none",scale="none",symm=F)
```
```{r, message=FALSE, warning=FALSE}
heatmap.2(input_heatmap,col=terrain.colors(100),Rowv=T,Colv=T,distfun = function(x) Dist(x, method="spearman"),dendrogram="both",key=T,ColSideColors=colorbar,density.info="none",trace="none",scale="none",symm=F)
```

# Step 12
> Produce a volcano plot and a Manhattan plot of the results of differential methylation analysis

For obtaining a Volcano plot, I need the difference in average methylation status for each probe among the 2 sample groups.
Since this is not specified in the assignment, I will use the normalized beta values from step 7 in this section, and the BH p-values from step 10.
Here I calculate the mean beta values for each probe in each sample group and their difference (across the 2 sample groups).


```{r, message=FALSE, warning=FALSE}
beta_norm_sorted <- complete_beta_p_df[,1:8]
beta_norm_sorted_DS <- beta_norm_sorted[,groups=="DS"]
beta_norm_sorted_WT <- beta_norm_sorted[,groups=="WT"]
mean_beta_DS <- apply(beta_norm_sorted_DS,1,mean)
mean_beta_WT <- apply(beta_norm_sorted_WT,1,mean)
delta <- mean_beta_DS - mean_beta_WT
summary(delta)
```

At this point `delta` contains the differential mean beta values between sample groups, for each position.
I can now plot the delta against the -log10 of the BH-corrected p-values.


```{r, message=FALSE, warning=FALSE}
neg_log10_p_BH <- -log10(complete_beta_p_df[,10])
plot(delta, neg_log10_p_BH,ylim=c(0,3),pch='.')
abline(a=-log10(0.05),b=0,col="red")
```
We can see how the p-values are binned at discrete levels (a consequence of the Mann-Whithney test) and nothing reaches the significance threshold of 0.05.

For obtaining the Manhattan plot, I first merge the dataframe containing the p-values for each probe with the annotation of the probes, using the `merge` function and merging according to probe name.


```{r, message=FALSE, warning=FALSE}
df_manhattan <- data.frame(rownames(complete_beta_p_df), complete_beta_p_df)
colnames(df_manhattan)[1] <- 'Name'
annotation <- getAnnotation(RGSet)
colnames(annotation)
complete_df_annotated <- merge(df_manhattan, annotation,by="Name")
colnames(complete_df_annotated)
```
Now I extract only the columns needed for obtaining the Manhattan plot (chromosome, position, BH p-value). I also rename the columns for clarity and order the levels according to the chromosome number.

```{r, message=FALSE, warning=FALSE}
chr_vect <- factor(complete_df_annotated$chr,levels=c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY"))
input_manhattan <- data.frame(chr_vect, complete_df_annotated$pos, complete_df_annotated$BH_p)
colnames(input_manhattan) <- c('chr','pos','BH_pval')
str(input_manhattan)
```
Finally I can obtain the Manhattan plot

```{r, message=FALSE, warning=FALSE}
palette <- rainbow(24)
mhtplot(input_manhattan,control=mht.control(colors=palette),ylim=c(0,3),pch=16)
axis(2,cex=0.5)
abline(a=-log10(0.05),b=0)
```
We can see how all the probes have a corrected p-value much below the significance threshold, and there are no significant differences across chromosomes.
We can also notice the binned distribution of the Mann-Whithney p-values.

Both the volcano plot and the Mabhattan plot have an appearance that is quite different from what I am used to with this dataset.
I suspect that this is due to the use of a non-parametric test (Mann-Whithney U test).
To test this hypothesis I will here repeat the analysis by using a Welch test and without changin the normalization procedure.

The code for the Welch test:
```{r, message=FALSE, warning=FALSE}
t_test_all_rows <- function(x) {
  t_val <- t.test(x~ targets$Group)
  return(t_val$p.value)
}

pval_raw_t <- apply(beta_norm,1, t_test_all_rows)
summary(pval_raw_t)
length(pval_raw_t)
```

The code for obtaining the joined dataset and for multiple testing correction:
```{r, message=FALSE, warning=FALSE}
beta_rawp_df_t <- data.frame(beta_norm, pval_raw_t)
beta_rawp_df_sorted_t <- beta_rawp_df_t[order(beta_rawp_df_t$pval_raw_t),]
summary(beta_rawp_df_sorted_t)
raw_p_sorted_t <- beta_rawp_df_sorted_t[,9]
summary(raw_p_sorted_t)
BH_p_t <- p.adjust(raw_p_sorted_t,"BH")
summary(BH_p)
Bonf_p_t <- p.adjust(raw_p_sorted_t,"bonferroni")
summary(Bonf_p_t)
complete_beta_p_df_t <- data.frame(beta_rawp_df_sorted_t, BH_p_t, Bonf_p_t)
```

The code for the volcano plot:
```{r, message=FALSE, warning=FALSE}
beta_norm_sorted_t <- complete_beta_p_df_t[,1:8]
beta_norm_sorted_DS_t <- beta_norm_sorted_t[,groups=="DS"]
beta_norm_sorted_WT_t <- beta_norm_sorted_t[,groups=="WT"]
mean_beta_DS_t <- apply(beta_norm_sorted_DS_t,1,mean)
mean_beta_WT_t <- apply(beta_norm_sorted_WT_t,1,mean)
delta_t <- mean_beta_DS_t - mean_beta_WT_t
summary(delta_t)
neg_log10_p_BH_t <- -log10(complete_beta_p_df_t[,10])
plot(delta_t, neg_log10_p_BH_t,ylim=c(0,3),pch='.')
abline(a=-log10(0.05),b=0,col="red")
```

The code for the Manhattan plot:
```{r, message=FALSE, warning=FALSE}
df_manhattan_t <- data.frame(rownames(complete_beta_p_df_t), complete_beta_p_df_t)
colnames(df_manhattan_t)[1] <- 'Name'
annotation <- getAnnotation(RGSet)
colnames(annotation)
complete_df_annotated_t <- merge(df_manhattan_t, annotation,by="Name")
colnames(complete_df_annotated_t)
chr_vect_t <- factor(complete_df_annotated_t$chr,levels=c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY"))
input_manhattan_t <- data.frame(chr_vect_t, complete_df_annotated_t$pos, complete_df_annotated_t$BH_p_t)
colnames(input_manhattan_t) <- c('chr','pos','BH_pval')
str(input_manhattan_t)
palette_t <- rainbow(24)
mhtplot(input_manhattan_t,control=mht.control(colors=palette_t),ylim=c(0,3),pch=16)
axis(2,cex=0.5)
abline(a=-log10(0.05),b=0)
```

Now the volcano and Manhattan plots look more familiar.

# Optional task
> As DS is caused by the trisomy of chromosome 21, try also to plot the density  of  the  methylation  values  of  the  probes  mapping  on chromosome 21.
> Do you see a very clear difference between the samples?
> How many differentially methylated probes do you find on chromosome 21?

First I extract the rows relative to probes belonging to chromosome 21 from the complete dataset prepared in step 12 and I subset it according to the group of the samples.

```{r, message=FALSE, warning=FALSE}
chr21_complete <- complete_df_annotated[complete_df_annotated$chr=='chr21',]
chr21_beta <- chr21_complete[,2:9]
chr21_beta_WT <- chr21_beta[,groups=='WT']
chr21_beta_DS <- chr21_beta[,groups=='DS']
dim(chr21_beta_DS)
dim(chr21_beta_WT)
```

Now I calculate the mean beta values of each probe in each of the sample groups, and the density of mean beta values.

```{r, message=FALSE, warning=FALSE}
chr21_beta_DS_mean <- apply(chr21_beta_DS,MARGIN=1,mean,na.rm=T)
chr21_beta_WT_mean <- apply(chr21_beta_WT,MARGIN=1,mean,na.rm=T)
d_mean_chr21_beta_DS <- density(chr21_beta_DS_mean)
d_mean_chr21_beta_WT <- density(chr21_beta_WT_mean)
```

Finally I plot the densities of mean beta values differentiated by sample group, only for probes on chromosome 21.

```{r, message=FALSE, warning=FALSE}
plot(d_mean_chr21_beta_DS,main="Density of Beta Values",col="red")
lines(d_mean_chr21_beta_WT,main="Density of Beta Values",col="blue")
```

Now I want to evaluate how many probes are differentially methylated on chromosome 21.
I will just extract this information from the complete data frame obtained in the previous steps.
I will use raw p-values here.

```{r, message=FALSE, warning=FALSE}
diff_met_chr21 <- chr21_complete[chr21_complete$pval_raw<=0.05,]
dim(diff_met_chr21)
```
On chromosome 21, 44 probes are differentially methylated according to raw p-value adopting a 0.05 threshold.
No significant p-values are present in the whole dataset after multiple testing correction, so there is no need to look for significant probes on chromosome 21.
